src/handlers/web3.ts

```diff
@@ -0,0 +1,679 @@
+  value?: TransactionRequest['value'];
+  from?: TransactionRequest['from'];
+  gasLimit?: string;
+  gasPrice?: string;
+}
+
+/**
+ * A type with the same keys as `NewTransaction` but without nullable types.
+ * This is useful for functions that assume that certain fields are not set
+ * to null on a `NewTransaction`.
+ */
+type NewTransactionNonNullable = {
+  [key in keyof NewTransaction]: NonNullable<NewTransaction[key]>;
+};
+
+/**
+ * @desc Configures `rpcEndpoints` based on a given `RainbowConfig`.
+ * @param config The `RainbowConfig` to use.
+ */
+export const setRpcEndpoints = (config: RainbowConfig) => {
+  rpcEndpoints[Network.mainnet] = config.ethereum_mainnet_rpc;
+  rpcEndpoints[Network.ropsten] = config.ethereum_ropsten_rpc;
+  rpcEndpoints[Network.kovan] = config.ethereum_kovan_rpc;
+  rpcEndpoints[Network.rinkeby] = config.ethereum_rinkeby_rpc;
+  rpcEndpoints[Network.optimism] = config.optimism_mainnet_rpc;
+  rpcEndpoints[Network.arbitrum] = config.arbitrum_mainnet_rpc;
+  rpcEndpoints[Network.polygon] = config.polygon_mainnet_rpc;
+};
+
+/**
+ * @desc web3 http instance
+ */
+export let web3Provider: JsonRpcProvider = (null as unknown) as JsonRpcProvider;
+
+/**
+ * @desc Checks whether or not a `Network | string` union type should be
+ * treated as a `Network` based on its prefix, as opposed to a `string` type.
+ * @param network The network to check.
+ * @returns A type predicate of `network is Network`.
+ */
+const isNetworkEnum = (network: Network | string): network is Network => {
+  return !network.startsWith('http://');
+};
+
+/**
+ * @desc set a different web3 provider
+ * @param {String} network
+ */
+export const web3SetHttpProvider = async (network: Network | string) => {
+  if (!isNetworkEnum(network)) {
+    web3Provider = new JsonRpcProvider(network, Network.mainnet);
+  } else {
+    web3Provider = new JsonRpcProvider(rpcEndpoints[network]);
+  }
+  return web3Provider.ready;
+};
+
+/**
+ * @desc returns true if the given network is a Layer 2
+ * @param network The network to check.
+ */
+export const isL2Network = (network: Network | string) => {
+  switch (network) {
+    case Network.arbitrum:
+    case Network.optimism:
+    case Network.polygon:
+      return true;
+    default:
+      return false;
+  }
+};
+
+/**
+ * @desc returns true if the given network is a testnet
+ * @param network The network to check.
+ */
+export const isTestnet = (network: Network) => {
+  switch (network) {
+    case Network.goerli:
+    case Network.kovan:
+    case Network.rinkeby:
+    case Network.ropsten:
+      return true;
+    default:
+      return false;
+  }
+};
+
+/**
+ * @desc returns a web3 provider for the specified network
+ * @param network The network as a `Network` or string.
+ */
+export const getProviderForNetwork = async (
+  network: Network | string = Network.mainnet
+) => {
+  if (networkProviders[network as Network]) {
+    return networkProviders[network as Network];
```

Comment:
Because the `networkProviders` object is a partial record, TypeScript thinks that the return type of this function can resolve to `Promise<undefined>` even though that would never happen, forcing consumers to handle it for no reason. I had a go at fixing it — here's my implementation:

```ts
export const getProviderForNetwork = async (
  network: Network | string = Network.mainnet
): Promise<JsonRpcProvider> => {
  if (isNetworkEnum(network) && networkProviders[network]) {
    return networkProviders[network] as JsonRpcProvider; // This ensures the return type doesn't resolve to undefined.
  }
  if (!isNetworkEnum(network)) {
    return new JsonRpcProvider(network, Network.mainnet);
  } else {
    const provider = new JsonRpcProvider(rpcEndpoints[network]);
    networkProviders[network] = provider;
    await provider.ready;
    return provider;
  }
};
```

Things to note:
 - I added an explicit `Promise<JsonRpcProvider>` return type to the function so that unexpected types like `undefined` can't slip through.
 - I added an `isNetworkEnum` call to the first `if` statement to make TypeScript happy since the `networkProviders` object can only be indexed by `Network` values.
 
 What do you think?

Reply:
This makes sense, thanks! I made those changes. I also integrated a change that got picked up while rebasing, which checks for `if (!networkProviders[network])` before updating `networkProviders` on line 158.